Unnest styles

difernandez opened the pull request:
## Context

- Banano is a component library for use in our projects
- A feature we're going to include is a `classes`, which is an object with keys that correspond to internal elements ofo a component, and allows to include new classes to those elements. This classes don't completely replace previous default styles, but should be able to override them if there is some overlap
- Developing this feature, we realized that there were some default styles that were added with selectors that had a higher specificity than a single class, so those styles were not being overriden

## Changes
This PR flattens all selectors of default styles to just be single classes, so that classes prop can override them.
- Every style applied with a nested selector, or a selector that checked more than one class, were changed to only need one class. This required some classes to be moved from the parent to the child
  - For listbox, a related fix was added: the disabled styles were being applied to the wrong element
  - Some tests that checked for error classes failed, I changed it to check for error class in any element within the component
- Color classes were also unnested. The previous behavior required a color class in the root element of the component, and a `colors:` key in the `.cjs`, in any sub element or modifier (or the block itself). Now the processing of these `colors:` are changed in the `tailwind/index.cjs`: it requires a color modifier class in the element itself. When using a `colors:` inside a modifier class, just one `-` is added, as BEM specifies that there should only be one modifier in a class name

difernandez committed:
feat(bn-checkbox): unnest styles

difernandez committed:
feat(bn-file-input): unnest styles

difernandez committed:
feat(bn-listbox): unnest styles and moved disabled styles to btn

difernandez committed:
feat(bn-textarea): unnest styles

difernandez committed:
fix(bn-textarea): bind attrs to textarea

difernandez committed:
feat(bn-toggle): unnest styles

difernandez committed:
feat(btn): unnest styles

difernandez commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Des-nestear las clases en los archivos cjs](https://trello.com/c/0F5VegsW/33-des-nestear-las-clases-en-los-archivos-cjs)

fprebolledo commented on src/components/BnCheckbox/BnCheckbox.styles.cjs:
```diff
@@ -6,9 +6,6 @@ module.exports = {
     },
     '&--error': {
```
This error key and belows are in the same nesting level. Maybe they can be on the same lvl.

difernandez replied:
Yo mean grouping both's styles in one `--error` key? They target two different elements, this one is for the whole component,  `bn-checkbox--error`, and the other one is just the input, the actual checkbox I mean, `bn-checkbox__input--error`, I don't think they can be grouped

fprebolledo replied:
aaah okay, I thought both of them were to the same component.

fprebolledo commented on src/components/BnCheckbox/BnCheckbox.styles.cjs:
```diff
@@ -17,6 +14,9 @@ module.exports = {
         '@apply text-varColor-600': {},
         '@apply focus-visible:border-varColor-600 focus-visible:ring-varColor-600': {},
       },
+      '&--error': {
```
This is the other key

fprebolledo commented on src/components/BnTextarea/BnTextarea.vue:
```diff
@@ -47,6 +48,7 @@ const attrs = useAttrs();
     :class="`bn-textarea--${props.color} ${attrs.class ? attrs.class : ''}`"
   >
     <textarea
+      v-bind="attrsWithoutClass"
```
Why in this component is it necessary?

difernandez replied:
because the attributes are applied by default on the root, the parent div in this case. So things like disabled="true" were not affecting the actual textarea.
This wasn't applied before because there was no parent div at some point, and in a PR I had to add it but I forgot this

fprebolledo replied:
I understand. I think you have to [disable inherited](https://vuejs.org/guide/components/attrs.html#disabling-attribute-inheritance) attrs to not apply it to the parent too.

fprebolledo replied:
I understand. I think you have to [disable inherited](https://vuejs.org/guide/components/attrs.html#disabling-attribute-inheritance) attrs to not apply it to the parent too.

difernandez replied:
uh you´re right! And I didn't know since 3.3 you could disable it in script setup 😮 nice

difernandez replied:
done!

fprebolledo commented on src/components/BnInput/BnInput.vue:
```diff
@@ -50,7 +50,7 @@ export default {
   <div
     class="bn-input"
-    :class="`bn-input--${props.color} ${attrs.class ? attrs.class : ''}`"
+    :class="attrs.class ? attrs.class : ''"
```
Maybe you could do this in this way: `:class = "{ attrs.class : attrs.class }"` instead

difernandez replied:
changed! (but github is having problems and it's not showing me the change)

fprebolledo commented on src/components/BnTextarea/BnTextarea.vue:
```diff
@@ -45,12 +45,15 @@ const attrsWithoutClass = Object.fromEntries(Object.entries(attrs).filter(([key]
   <div
     class="bn-textarea"
-    :class="`bn-textarea--${props.color} ${attrs.class ? attrs.class : ''}`"
+    :class="attrs.class ? attrs.class : ''"
```
Maybe you could use hash notation instead ternary conditional

fprebolledo reviewed (approved):
It looks good. I left a comments about how have being passed classes to a component.

difernandez committed:
feat(tailwind): change how color classes are parsed

Now it expects a color class variant for each element that has varColor related changes

difernandez committed:
feat(yarn): upgrade vue to 3.3

difernandez committed:
fix(bn-textarea): add missing inheritAttrs: false
